### ISSUE TITLE
Add tiered, monthly subscription pricing and payment

### ISSUE THREAD

**Issue by BenJam** [mysociety/mapit.mysociety.org#26]
Users should be able to sign up for an account, purchase the correct subscription (inc. 50% discount for registered charities), set this subscription to automatically renew and manage their subscriptions online.

**Comment by abibroom** [comment]
As a user, I need to be able to

* See/change my details
  * Name
  - [x] Email address
  - [x] Password
  * Company name
  * VAT number
  * Billing address
- [x] Add or disable API keys
- [x] See my keys
- [x] See my usage per key,
- [x] and in total (including disabled keys?)
- [x] See what plan I'm on
- [ ] Be alerted by email when I'm close to the usage limit on my plan (requires #25)
- [x] Compare plans
- [x] Upgrade or downgrade my plan
- [x] Request to be put on the special charities plan (#24)
- [x] See how much my next bill will be, and when the payment will be taken
- [x] Receive an email receipt when a payment is successfully taken (from Stripe)
- [ ] Receive an email when a payment fails
- [ ] ~~Receive an email when a payment card is about to expire~~
- [x] See/change my payment card details
- [ ] View my billing history in summary - 1st Jan paid £X, 1st Feb paid £X, etc
- [ ] View my billing history in detail - for each previous payment, an invoice showing:
  * mySociety's details
  * details of service paid for, including dates
  * amount, with VAT itemised separately
  * my billing address
  * my VAT number

Enhancements:

* Add a second email address (finance department) to receive a copy of the email when a payment is successfully taken or fails
* Download invoices as PDFs
* Add a secondary user(s) to my account
* Transfer account 'ownership' (the billing) to one of those other users instead of me

**Comment by stevenday** [comment]
My idea of an MVP for this would be:

- [x] Form/forms on your account to pay for an annual subscription, where you select the plan you want and enter any/all the details we need to charge you like card details, address, VAT number, etc. This *could* do some kind of auto-renewal, but we could also try to code it in such a way that we save the details we need (a token from Stripe for example) and work out the actual repeat charging details later - we'll have at least a year. We can drop you into this page after you've created an account so that it's part of the workflow, but also make it available from the user account menu.
- [x] Receive an email receipt when payment is successfully taken (~~alas Stripe don't do this automatically so we will have to code this~~)
- [x] See what plan I'm on (on user account page)
- [x] Compare plans (on a marketing page and again on the payment page to choose which one to buy)
- [x] ~Add or disable API keys~ (you can already do this on the account page)
- [x] ~See my keys~ (you can already do this on the account page)
- [x] See my usage per key,
- [x] and in total (including disabled keys) (on user account page, needs #25)

Everything else could be initially done via email (upgrade or downgrade your account, get charity account) or is an enhancement (recurring billing, extra emails for billing, usage alerts, billing history).

**Comment by abibroom** [comment]
Worth bearing in mind that our top level annual charges, £3000+VAT for example, may be beyond spending limits on debit/credit cards. (Also psychologically, £300/month sounds like less money than £3600/year.)

**Comment by stevenday** [comment]
Ok, so from discussion with @abimysoc and @markcridge, monthly billing is a must have, which complicates my MVP a little bit. I think [stripe's subscription stuff](https://stripe.com/subscriptions) looks like the best fit for this, as it handles the basics plus more advanced stuff like trial periods, add-ons, discounts etc. I went through their docs to figure out how it might be done as simply as possible and have sketched it out below.

Instead of a single payment page where we take annual subscriptions effectively as a one-off payment, we're going to need to have a bit more stuff to allow people to actually manage their account:

- [x] A page to create a subscription in the first place, choosing the plan and entering payment details (this can use [Stripe's checkout.js](https://stripe.com/docs/recipes/subscription-signup))
- [x] A similar page to update the stored card details (You can [use checkout.js for this too](https://stripe.com/docs/recipes/updating-customer-cards))
- [x] In addition, we'll need to register a url to receive Stripe's webhooks, so that we can act upon things that might happen, both good stuff like invoices being paid and bad stuff like declined payments, cards expiring (though Stripe [has some magic to resolve that apparently](https://stripe.com/blog/smarter-saved-cards)), subscriptions being cancelled, etc. 
- ~~Stripe doesn't do any emailing of the customer for us, so we'll need to add all of those appropriate emails too~~

As with all payment stuff, everything is a bit asynchronous, so we need to decide what we're going to do when there isn't an appropriate payment. Do we be nice and let people use the API anyway (in which case we have to have some secondary system to disable accounts eventually), or do we be mean and make people wait until their payment has cleared/cut them off immediately? 

I'd lean towards the former, to avoid the need to send reminders etc beforehand, with some kind of manual process (I think we could just look in the Stripe Dashboard) to find serious offenders.

I still think that upgrading/downgrading/cancelling can be left as enhancements for now, until we see that they're a big enough time sink to justify building them in code?

**Comment by abibroom** [comment]
> what we're going to do when there isn't an appropriate payment

My expectation as a user of subscription-based services would be for things to carry on working, possibly with functionality limited in some way, definitely with a big red warning banner everywhere ("you need to [action] by [date] or [bad thing will happen]"). On the admin side we'd need a means of being aware of when a bad thing needs to happen, which could be as simple as a monthly reminder in someone's calendar to go and check for non-paying accounts.

> I still think that upgrading/downgrading/cancelling can be left as enhancements for now

Probably? Users can effectively achieve cancellation by not paying, if we do the above. Upgrades and downgrades are rare at the moment (but if it becomes easier to see your usage and there's something in place to stop you exceeding a limit, we may see it becoming a more common requirement).

**Comment by dracos** [comment]
Okay, so here's the to do list as I see it to have normal signup working:
- [x] Create plans (mapit-5k/10k/20k/30k/50k) and coupons (charitable50 and charitable100) in Stripe.
- [x] Create a Subscription model, that stores the User (ie. email), charitable status/justification, and their subscription ID (from stripe). We can get the Stripe customer ID, current period start/end, discount applied, etc, from Stripe, we don't need to store it ourselves.
- [x] Adapt the signup view so that it asks for the above (plus one separate for already existing people), where you can pick a plan, agree to T&Cs, say whether you're charity (provide number)/non-profit unpaid individual (provide justification)/other charitable usage and then pay for it, with Stripe. This creates a Subscription object.
- [x] Create a current subscription view page that outlines current usage, end date, amount to be taken next, etc.

Stripe says it does send success receipt emails, so I think I can avoid any webhook work for the moment. We can manually look at the dashboard for failures (they might email us about them too).

Things that are then nice to have but can also be done manually in Stripe dashboard:
- [x] Update payment details (note this is not necessary for normal expiry/card changes as Steve mentioned above)
- [x] Upgrade/downgrade (I think is very simple in that you call Stripe's update subscription API call and it works out the pro-rata stuff for you for the next invoice automatically)

**Comment by stevenday** [comment]
> Stripe says it does send success receipt emails, so I think I can avoid any webhook work for the moment

Oh, so it does. I think I've even set up a logo and stuff on the mySociety Stripe account already so that these look halfway decent. However, the issue I remember now is that these only include a single total amount, there's nowhere to specify VAT (because you don't tell Stripe about VAT generally, you're supposed to calculate it and then just charge the lump sum). 

I don't know if we could get away with hacking the VAT amount into some of the text we send to Stripe, e.g. in the item description?

Otherwise, sounds good. I think I might name Plan -> Subscription but whatever. (A plan is the payment level you choose to me, where as a subscription covers all the stuff you're talking about storing like customer details, etc).

**Comment by dracos** [comment]
Model rename - true, I did even say it would store a subscription ID, didn't I :)

**Comment by dracos** [comment]
The PR at #32 now includes plan picking/payment as part of signup. You can also add a subscription to an existing account (for those people already with accounts), and you can upgrade/downgrade your plan at any time using the same form (it should work out whether it needs to ask you for payment details or not hopefully!). There is a view page that lists your current plan, payment/card, creation date and expiry (though it will auto-renew, so expiry might not be the best word, unless/until we had a cancel option – perhaps drop that display line). Still need to add T&Cs tickbox agreement.

<img src="https://cloud.githubusercontent.com/assets/154364/22258327/0f7972fe-e259-11e6-853d-b922985c439b.gif" width="50%">

**Comment by dracos** [comment]
So, I think tech wise this is all basically there. If they hit their usage limit all their keys are blocked, if they upgrade their plan they'll get unblocked, or wait until next successful payment from Stripe. If they don't pay… Currently nothing happens (well, their quota won't be reset so it'll presumably get hit at some point), and if you manually then cancel their subscription in the Stripe dashboard, they'll go back to no-subscription-at-all tier. I think that works?

Todo:
- [x] Show per key usage
- [x] Push previous month usages onto a redis list
- [x] Script to run at midnight and unblock daily-blocked IP addresses
- [x] Send emails on receiving unpaid invoice notification
- [x] T&C tickbox
- [x] EMails need testing with fake failed invoice through Stripe, see https://stripe.com/docs/recipes/sending-emails-for-failed-payments

Anything else?

**Comment by abibroom** [comment]
Design note: maybe a totaliser like on https://www.theyworkforyou.com/video/ to show usage?

**Comment by markcridge** [comment]
Is there any manual override if we wish to grant some organisation free
access for any reason, or extend a subscription at a lower level?

**Comment by dracos** [comment]
Yes; you can e.g. give a subscription a coupon of any value, including 100%. Or create a custom plan for them.
